### PR TITLE
 [router] Add gRPC E2E test suite

### PR DESCRIPTION
**Pull request by key4ng**
## Motivation
The SGLang router supports gRPC connections but lacks end-to-end tests to validate this functionality. This PR adds comprehensive test coverage for gRPC endpoints to ensure reliability and catch regressions.

## Modifications

- Added 14 new test files under sgl-router/py_test/e2e_grpc/ 
  - Test categories:
    - Basic endpoints: /v1/models, /v1/chat/completions
    - Embedding functionality
    - Function calling with tool choice
    - Advanced features: cache reporting, EBNF grammar, hidden states
    - Edge case validation (e.g., large max_new_tokens)
  - Test structure: Organized into basic/, function_call/, features/, and validation/ subdirectories
## Accuracy Tests
https://github.com/sgl-project/sglang/actions/runs/18694641875/job/53308814787?pr=11790

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: sgl-router/py_test/e2e_grpc/features/test_openai_server_ebnf.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+"""
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/features/test_openai_server_ebnf.py]: We can remove this one, it is not needed in grpc

File: sgl-router/py_test/e2e_grpc/function_call/test_tool_choice.py [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+# CHANGE: Import router launcher instead of server launcher
+import sys
+from pathlib import Path
+_TEST_DIR = Path(__file__).parent
+sys.path.insert(0, str(_TEST_DIR.parent))
+from fixtures import popen_launch_grpc_router
+
+import openai
+
+from sglang.srt.utils import kill_process_tree
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/function_call/test_tool_choice.py]: can we remove deps from sglang.srt?

File: sgl-router/py_test/e2e_grpc/basic/test_openai_embedding.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+"""
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/basic/test_openai_embedding.py]: embedding is not implemented.

File: sgl-router/py_test/e2e_grpc/features/test_openai_server_hidden_states.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
+"""
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/features/test_openai_server_hidden_states.py]: Same, this one is not implemented yet

File: sgl-router/py_test/e2e_grpc/validation/test_large_max_new_tokens.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    STDOUT_FILENAME,
+    CustomTestCase,
+
+)
+
+class TestLargeMaxNewTokens(CustomTestCase):
+    @classmethod
+    def setUpClass(cls):
+        # CHANGE: Launch gRPC router with integrated workers (single command)
+        cls.model = "/home/ubuntu/models/llama-3.1-8b-instruct"
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/validation/test_large_max_new_tokens.py]: typo?

File: sgl-router/py_test/e2e_grpc/features/test_cache_report.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
+"""
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/features/test_cache_report.py]: I am not sure how this test works.

File: sgl-router/py_test/e2e_grpc/basic/test_openai_server.py [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+
+        if use_list_input:
+            prompt_arg = [prompt_input, prompt_input]
+            num_choices = len(prompt_arg)
+            num_prompt_tokens *= 2
+        else:
+            prompt_arg = prompt_input
+            num_choices = 1
+
+        generator = client.completions.create(
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/basic/test_openai_server.py]: completions is not supported yet.

**Review by CatherineSue — COMMENTED** [review]
Can we simplify this by verifying the /v1/models and /v1/chat/completions endpoint first? 

I think for advanced features, we don't need cache_report, ebnf, hidden_states yet.

File: sgl-router/py_test/e2e_grpc/features/test_cache_report.py (hunk 6 above)
Comment (key4ng) [reply sgl-router/py_test/e2e_grpc/features/test_cache_report.py]: removed

File: sgl-router/py_test/e2e_grpc/features/test_reasoning_content.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            "max_tokens": 100,
+            "extra_body": {"separate_reasoning": True},
+        }
+        response = client.chat.completions.create(**payload)
+
+        assert len(response.choices[0].message.reasoning_content) > 0
+        assert len(response.choices[0].message.content) > 0
+
+
+class TestReasoningContentWithoutParser(CustomTestCase):
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/features/test_reasoning_content.py]: can we remove this test or rename it?

Because it is basically duplicated to `test_nonstreaming_separate_reasoning_false` above.

File: sgl-router/py_test/e2e_grpc/features/test_enable_thinking.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+import unittest
+
+# CHANGE: Import router launcher instead of server launcher
+from pathlib import Path
+
+import openai
+import requests
+
+_TEST_DIR = Path(__file__).parent
+sys.path.insert(0, str(_TEST_DIR.parent))
```
Comment (CatherineSue) [inline sgl-router/py_test/e2e_grpc/features/test_enable_thinking.py:23]: why do we need this for all uts?

File: sgl-router/py_test/e2e_grpc/features/test_reasoning_content.py (hunk 8 above)
Comment (key4ng) [reply sgl-router/py_test/e2e_grpc/features/test_reasoning_content.py]: sure removed

File: sgl-router/py_test/e2e_grpc/features/test_enable_thinking.py (hunk 9 above)
Comment (key4ng) [reply sgl-router/py_test/e2e_grpc/features/test_enable_thinking.py:23]: This adds the e2e_grpc directory to Python's module search path, allowing from util import ... to work

File: .github/workflows/pr-test-rust.yml [hunk 10]
```diff
@@ -144,6 +144,12 @@ jobs:
           python3 -m pip --no-cache-dir install --upgrade --break-system-packages genai-bench==0.0.2
           pytest -m e2e -s  -vv -o log_cli=true --log-cli-level=INFO
 
+      - name: Run Python E2E gRPC tests
```
Comment (slin1237) [inline .github/workflows/pr-test-rust.yml:147]: shall we use another workflow?
to speed things up

Comment (key4ng) [reply .github/workflows/pr-test-rust.yml:147]: yeah good point. we can do that later once the openai router ci is also added
